sycl/source/detail/global_handler.cpp

```diff
@@ -150,7 +150,8 @@ extern "C" __SYCL_EXPORT BOOL WINAPI DllMain(HINSTANCE hinstDLL,
   // Perform actions based on the reason for calling.
   switch (fdwReason) {
   case DLL_PROCESS_DETACH:
-    shutdown();
+    if (!lpReserved)
+      shutdown();
```

Comment:
Are we 100% positive, there's no problem with GPUs and FPGAs, if we don't call all the necessary release APIs upon exiting from the process? Definitely contributes to analysis tools reports.

cc @romanovvlad @jandres742 @pwilma @GarveyJoe

Reply:
I refer to [this](https://docs.microsoft.com/en-us/windows/win32/dlls/dllmain) manual and [blog post](https://devblogs.microsoft.com/oldnewthing/20120105-00/?p=8683). The manual says:
> When handling DLL_PROCESS_DETACH, a DLL should free resources such as heap memory only if the DLL is being unloaded dynamically (the lpReserved parameter is NULL). If the process is terminating (the lpvReserved parameter is non-NULL), all threads in the process except the current thread either have exited already or have been explicitly terminated by a call to the ExitProcess function, which might leave some process resources such as heaps in an inconsistent state. In this case, it is not safe for the DLL to clean up the resources. Instead, the DLL should allow the operating system to reclaim the memory.

The second sentence explicitly says, that current thread is the last one working and is not safe to perform calls to other DLLs as those might already received their `DLL_PROCESS_DETACH` notification.

Reply:
@s-kanaev that's true, but I presume, that blog post only refers to file descriptors and memory allocations. Do we have the same guarantees for other kernel objects?

Reply:
No, I don't have such guarantees.

Reply:
Is the question: "does anything bad happen in the FPGA OpenCL runtime if clReleaseContext is not called before program termination"?  If so, @pcolberg can probably answer.

Reply:
From what I know, the memory kept track in platform and all context will be leaked until process is terminated. If the program runs in multi-threaded fashion and all memory dangle like this until process end, then it might impose a limitation on how much thread user can spawn. (its became just like CUDA)

Reply:
@sherry-yuan do I read this right? If we don't call clReleaseContext before exiting the process, and this process terminates, FPGA board will be locked until someone reboots the system?

@jandres742 @pwilma does anything bad happen to GPUs if we don't release them properly?

Reply:
by release you mean releasing/destroying all the objects you have created? if so, there are no problems, as at the end the OS would kill everything associated with the process one the process finishes.

Reply:
> assuming device is closed on process exit

@sherry-yuan , could you, please, elaborate what does it mean that device is closed?

Reply:
@s-kanaev Short answer: when device is closed, then it become accessible to other process. Only one process can open a device at a time.

Each context occupies a device (open up a device). When a device is open by one process, it cannot be opened by another process. a part of clReleaseContext's job is to close the device so that it become accessible to other process before current process terminates. @alexbatashev I believe the device is closed regardless when current process exit, correct me if I am wrong @GarveyJoe

Reply:
@jandres742 Right, that's what I mean by "context will be leaked until process is terminated". AFAIK The limitation I can see from this is the max number of thread a user can spawn due to memory limitation.

Previously, some memory will be cleared up if a thread finishes early. but now they will occupy the memory until process ends.

Reply:
@sherry-yuan  correct. Main issue here is memory leaking and that wont be free until process exists. Now, are we trying to do this always or only when something happens? if we are trying to do it always, then it seems a bit weird leaving that mem leaking  and not cleaning properly. why is that we need to do this?

Reply:
@sherry-yuan , I still can't get how one closes a device? How is it performed from DPCPP RT point of view? Is it releasing of device/context or anything else?

Reply:
@s-kanaev I was replying to Joe's question regarding "what happens if clReleaseContext is not called before program termination".

To answer your question, device is closed when clReleaseContext is called, and that was called when shutdown() is called
https://github.com/intel/llvm/blob/e386efe2e8a662f7efd64034cb0acaa2a4a16118/sycl/source/detail/global_handler.cpp#L154
Then shutdown calls on default context's release which has the intentional leak, i.e does not properly call the destructor of context_impl
https://github.com/intel/llvm/blob/e386efe2e8a662f7efd64034cb0acaa2a4a16118/sycl/source/detail/global_handler.cpp#L97-L102

If that specific lines does exactly the same as what linux does (like below)
https://github.com/intel/llvm/blob/e386efe2e8a662f7efd64034cb0acaa2a4a16118/sycl/source/detail/global_handler.cpp#L95

Then the destructor of context_impl will call on clReleaseContext in OpenCL runtime, and that function closes the device.

Now this function is not called on windows, i.e the device associated with the contexts is not closed until process exit.

I am not sure I am answering the right question, this problem seems pre-existing.

Reply:
Right, then, it's nothing but memory leak.